Topics/Tech_Stacks/build-android-app-using-github-actions.md

```diff
@@ -0,0 +1,158 @@
+
+### Build steps
+
+```yml
+    # Optional pre-build step
+    # Make sure to cd to your android folder though!
+    - name: Prebuild Android
+      working-directory: Frontend/DebrisMine
+      run: |
+        npx expo prebuild -p android --no-install
+        cd android && chmod +x ./gradlew
+
+    # Building using gradle
+    - name: Build Android Release
+      working-directory: Frontend/DebrisMine
+      run: |
+        cd android && ./gradlew clean && ./gradlew app:assembleRelease
+```
+
+Since I use expo, I have a prebuild step before building which generates the android folder for me. If you are not using expo, then simply ommit this line `npx expo prebuild -p android --no-install`, and just cd into the android folder which you should have pushed to your repo.
+
+### Signing steps
+
+All Android apps need to be signed for them to be trusted by your Android device (and Google Play).
+
+```yml
+    # Sets the environment variable BUILD_TOOL_VERSION based on the installed
+    # Android SDK version (needed for next step).
+    - name: Setup build tool version variable
+      shell: bash
+      run: |
+        BUILD_TOOL_VERSION=$(ls /usr/local/lib/android/sdk/build-tools/ | tail -n 1)
+        echo "BUILD_TOOL_VERSION=$BUILD_TOOL_VERSION" >> $GITHUB_ENV
+        echo Last build tool version is: $BUILD_TOOL_VERSION
+
+    # Sign the APK
+    - name: Sign APK
+      id: sign_app
+      uses: r0adkll/sign-android-release@v1
+      with:
+        releaseDirectory: ${{ github.workspace }}/Frontend/DebrisMine/android/app/build/outputs/apk/release
+        signingKeyBase64: ${{ secrets.ANDROID_SIGNING_KEY }}
+        alias: ${{ secrets.ANDROID_SIGNING_ALIAS }}
+        keyStorePassword: ${{ secrets.ANDROID_SIGNING_STORE_PASSWORD }}
+        keyPassword: ${{ secrets.ANDROID_SIGNING_KEY_PASSWORD }}
+      env:
+        BUILD_TOOLS_VERSION: ${{ env.BUILD_TOOL_VERSION }}
+```
+
+To sign the APK you will need to add four secrets to your GitHub repo:
+`ANDROID_SIGNING_KEY`, `ANDROID_SIGNING_ALIAS`,
+`ANDROID_SIGNING_STORE_PASSWORD`, and `ANDROID_SIGNING_KEY_PASSWORD`.
+
+[Keytool](https://docs.oracle.com/javase/7/docs/technotes/tools/solaris/keytool.html) comes installed with the JDK. You can create a key using Keytool from the command line: `keytool -genkey -v -keystore my-release-key.keystore -alias alias_name -keyalg RSA -keysize 2048 -validity 10000`. You can replace alias_name with anything you'd like, but make sure to remember it, since it will be the `ANDROID_SIGNING_ALIAS` secret. The command line tool will aslo ask for a store password and a key password (make sure to remember it or note it down somewhere). Finally, the `ANDROID_SIGNING_KEY` is simply the contents of the generated file: `my-release-key.keystore`.
+
+![](./assets/actions-secrets.png)
+You can add the secrets through the repo settings. 
+
+### Upload step
+
+The app has been build but it's still "stuck" on the machine, we need to upload
+it back to GitHub as an Artifact.
```

Comment:
Could benefit the reader to link github artifacts here: https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts

Reply:
done